llvm/lib/SYCLLowerIR/SYCLUtils.cpp

```diff
@@ -43,6 +47,13 @@ void traverseCallgraphUp(llvm::Function *F, CallGraphNodeAction ActionF,
         } else {
           // ... non-call is OK - add using function to the worklist
           if (auto *I = dyn_cast<Instruction>(FCall)) {
+            if (auto *SI = dyn_cast<StoreInst>(I)) {
```

Comment:
The StoreInst check should be moved into the filter as it has nothing to do with this generic traversal functionality. That's why I suggested `std::function<bool(const Instruction *I, const Function *F)>`, not `std::function<bool(const Value *)>`
```suggestion
            if (!functionFilter(I, CurF)) {
                continue;
            }
```